Styling support

X-m7 opened the pull request:
@sean0x42 Can you please have a look at the styling structure in demo.json? Turns out that React can take a JS object with a mapping of CSS properties to values and turn it into a string for the browser to read, I thought it'd be more complicated. So, if that's all good I'll go ahead and extend the ODT parser to support it.

## Currently supported features
 - Bold
 - Italic
 - Underline (including colour and style in the parsing side)
 - Font family
 - Font size (only supports the CSS absolute font sizes in the interface due to a limitation in `Document.execCommand()`)
 - Font colour

## Proposed Changes
 - Add styles as part of our DOM (currently implemented with inline styles and span tags where needed)
 - Add support for the client to render the styles from the DOM
 - Add some styling examples to the demo.json
 - Add buttons/menus in the client for common styling features
 - Add support to parse styles in the ODT parser

## Related Stories
 - [Add styling support](https://documentprocessor.atlassian.net/browse/DOCS-82)

## Acknowledgements
 - [x] I have read and followed the [Contributing guide](https://github.com/sean0x42/kauri/blob/master/.github/CONTRIBUTING.md).
 - [x] I have followed the branching requirements.
 - [x] I have run `rustfmt` and `yarn clean` on all proposed Rust and JavaScript.

sean0x42 commented:
Looks good to me. Only thing I can see is that the convention is to use camel case I believe. So where you have `text-decoration` for example I believe it should be `textDecoration`. This is exactly what I was planning on doing anyway so go for it! Great minds think alike 😛

X-m7 commented:
So I've just partially implemented the ODT parsing side, and now I'm getting a warning from [Clippy](https://github.com/rust-lang/rust-clippy) (the official Rust linter, which Eclipse uses by default, so now it underlines the whole thing in yellow) that the function is probably getting too big (cyclomatic complexity of 28). So, I guess it would be a good time to split the function up. @sean0x42 Would it be fine for me to do that in this pull request or should it be done in another one?

sean0x42 commented:
That's fine @X-m7, go ahead and do it here if you like. Ideally I'd like there to be a different function for each tag type

X-m7 commented:
While working on this I ended up finding an inconsistency between ODT styling and CSS regarding underline color when it is set to follow the text color. Essentially, if we have an outer block of text with underline styling which has an inner block of text with text color styling, the inner block's underline color will have the same color as the rest, while in LibreOffice the inner block's underline will follow its own color. I think we can handle this by adding another underline style just for the inner block, but that would require us to check if any of the parents have underline color set to currentcolor and that the inner block color is set differently to that parent's color (probably?). Is this case worth handling at this stage or should I just continue implementing other stuff? It is easy enough to be fixed by the user (just set the underline style explicitly after import), at least if it's not used extensively.
![Screenshot_20190506_165832](https://user-images.githubusercontent.com/18729369/57210827-7626d980-7021-11e9-8ec1-78c85d98228a.png)

sean0x42 commented:
Turns out this functionality is supported by CSS if you set `text-decoration: underline` on the child element as well (without explicitly declaring the underline colour). So that might reduce the overhead a little since you don't need to compare the parent/child elements underline colour.

![T](https://i.imgur.com/OQBbfsB.png)

I think it's a good idea to try and support everything from ODT as transparently as possible, so if it's not too much of a hassle I say give it a shot.

Also nice work so far!

X-m7 commented on client/src/components/Editor/ToolBar/styles.scss:
```diff
@@ -1 +1,3 @@
-
+.toolbar * {
+    display: inline-block;
```
Fixed by commit 16114c3

X-m7 commented:
@sean0x42 What should the target list of features be for this pull request? At this point I can think of strikethroughs, highlights, superscripts, subscripts, alignment, lists, tables, as well as named styles (the default heading and paragraph styles for example), but then adding all of them would result in an even bigger pull request (Apparently it's XL now, although a large chunk of it is actually yarn.lock), so I'm not sure if it would be better to put those in a separate pull request so as to make reviewing this one less painful. Anyway, I think it's decent enough for demonstration purposes during the meeting with Alex on Wednesday.

Speaking of named styles, I'm thinking that we can just turn them into CSS classes, but then what's the best way to generate CSS classes in the client?

sean0x42 commented:
> (Apparently it's XL now, although a large chunk of it is actually yarn.lock), so I'm not sure if it would be better to put those in a separate pull request so as to make reviewing this one less painful.

Yes I 100% agree. I think it'd be a good idea to split this into multiple focused PRs from here out.

> as well as named styles (the default heading and paragraph styles for example) [...] Speaking of named styles, I'm thinking that we can just turn them into CSS classes, but then what's the best way to generate CSS classes in the client?

Named styles are an interesting beast. CSS classes make sense when exporting to HTML for example, but not so much when we're trying to render them, because we cannot change the class definitions dynamically. Leave them out for now, we'll have to discuss them a bit further at our next meeting I think.

> Anyway, I think it's decent enough for demonstration purposes during the meeting with Alex on Wednesday.

That's good to hear. As far as I know, Alex hasn't made a formal statement about when all the groups are meeting him so I might send him an email.

X-m7 commented:
Okay, so I'll take this out of draft then, and I'll start working on the other features once this is merged.

X-m7 commented on client/main.js:
```diff
@@ -32,3 +33,18 @@ app.on("activate", () => {
+
+ipcMain.on("getFontList", (event, args) => {
+  var fontArray = [];
+  var fonts = fontManager.getAvailableFontsSync(); //there is an async version that takes a function, not sure how to use that here
```
The thing I wasn't sure about was how to actually send the response from within the callback, but now that I think about it that's obvious, clearly I still need to get used to functional programming and JS in general. Anyway, fixed by 714e495.

X-m7 commented on client/main.js:
```diff
@@ -32,3 +33,18 @@ app.on("activate", () => {
+  for (var i = 0; i < fonts.length; i++) {
+    fontArray.push(fonts[i].family);
+  }
+  var uniqueFontArray = fontArray.filter(onlyUnique);
```
Fixed by 714e495.

X-m7 commented on client/main.js:
```diff
@@ -32,3 +33,18 @@ app.on("activate", () => {
+  var fonts = fontManager.getAvailableFontsSync(); //there is an async version that takes a function, not sure how to use that here
+  for (var i = 0; i < fonts.length; i++) {
+    fontArray.push(fonts[i].family);
+  }
```
Fixed by 714e495.

X-m7 commented on client/main.js:
```diff
@@ -32,3 +33,18 @@ app.on("activate", () => {
+}
+
+ipcMain.on("getFontList", (event, args) => {
+  var fontArray = [];
```
Fixed by 714e495.

sean0x42 commented on client/src/dom/render.jsx:
```diff
@@ -32,11 +32,15 @@ export function renderDocumentNodes(nodes) {
   switch (node.type) {
     case "heading":
-      return <Heading node={node} />;
+      return <Heading node={node} style={node.style} />;
```
Doesn't this result in sort of duplication of the style object? i.e. A heading instance could access it's styles via either `props.node.style` or `props.style`?

X-m7 replied:
Nice catch, didn't realize it was already handling the style by itself, fixed by f5557d8

X-m7 commented on client/src/components/Editor/ToolBar/FontSelector.jsx:
```diff
@@ -0,0 +1,44 @@
+    ipcRenderer.send("getFontList");
+    ipcRenderer.on("fontList", (event, args) => {
+      this.setState({ fontList: args });
+    });
```
Fixed by b27531b

X-m7 commented on client/src/components/Editor/ToolBar/FontSelector.jsx:
```diff
@@ -0,0 +1,44 @@
+    var options = [];
+    for (var i = 0; i < this.state.fontList.length; i++) {
+      options.push(<option>{this.state.fontList[i]}</option>);
+    }
```
Fixed by f5557d8

X-m7 commented on client/src/components/Editor/ToolBar/FontSizeSelector.jsx:
```diff
@@ -0,0 +1,34 @@
+    var options = [];
+    for (var i = 1; i <= 7; i++) {
+      options.push(<option>{i}</option>);
+    }
```
Fixed by f5557d8

X-m7 commented:
I just merged the changes from #12 in, also added in your suggestions and switched the font dependency to one that doesn't include bogus fonts (which don't do anything when selected, I guess those are like bitmap fonts or something).

sean0x42 commented on client/src/main/index.js:
```diff
@@ -48,3 +50,15 @@ app.on("activate", () => {
 });
+
+ipcMain.on("getFontList", (event, args) => {
+  systemFonts.getFonts().then(
```
Nice! Promises fit much better here 👍
